hknweb/candidate/views.py

```diff
@@ -21,50 +48,87 @@ class CandRequestView(FormView, generic.ListView):
+                # TODO: for some usernames such as catherine.hu, this becomes a link. Why??
+                'candidate_username' : form.instance.requester.username,
+                'confirm_link' : confirm_link,
+                'img_link' : get_rand_photo(),
             }
         )
         msg = EmailMultiAlternatives(subject, text_content,
                 'no-reply@hkn.eecs.berkeley.edu', [officer_email])
         msg.attach_alternative(html_content, "text/html")
         msg.send()
 
-    def get_context_data(self, **kwargs):
-        context = super(CandRequestView, self).get_context_data(**kwargs)
-        return context
-
     def get_queryset(self):
-        result = OffChallenge.objects
+        result = OffChallenge.objects \
+                .filter(requester__exact=self.request.user) \
+                .order_by('-request_date')
+        return result
+
+
+# List of past challenge requests for officer
+# Non-officers can still visit this page but it will not have any entries
+@method_decorator(login_required(login_url='/accounts/login/'), name='dispatch')
+class OffRequestView(generic.ListView):
+    template_name = 'candidate/offreq.html'
+
+    context_object_name = 'challenge_list'
 
-        result = result.order_by('-request_date').filter(requester=self.request.user)
+    def get_queryset(self):
+        result = OffChallenge.objects \
+                .filter(officer__exact=self.request.user) \
+                .order_by('-request_date')
         return result
 
 
+# Officer views and confirms a challenge request after clicking email link
+# Only the officer who game the challenge can review it
+@login_required(login_url='/accounts/login/')
 def officer_confirm_view(request, pk):
+    def send_cand_confirm_email(form):
+        subject = 'Your Officer Challenge Was Reviewed'
+        candidate_email = form.instance.requester.email
+        text_content = 'Your Officer Challenge Was Reviewed'
+
+        challenge_link = request.build_absolute_uri(
+                reverse("candidate:detail", kwargs={ 'pk' : form.instance.id }))
+        html_content = render_to_string(
+            'candidate/cand_confirm_email.html',
+            {
+                'confirmed' : form.instance.confirmed,
+                'officer_name' : form.instance.officer.get_full_name(),
+                'officer_username' : form.instance.officer.username,
+                'challenge_link' : challenge_link,
+                'img_link' : get_rand_photo(),
+            }
+        )
+        msg = EmailMultiAlternatives(subject, text_content,
+                'no-reply@hkn.eecs.berkeley.edu', [candidate_email])
+        msg.attach_alternative(html_content, "text/html")
+        msg.send()
+
+    # TODO: gracefully handle when a challenge does not exist
     challenge = OffChallenge.objects.get(id=pk)
+    if request.user.id != challenge.officer.id:
+        return render(request, "candidate/401.html", status=401)
```

Comment:
Dovetails with the above comment, of using a generic 401 page than a candidate-specific 401 page.